text/src/od.rs

```diff
@@ -0,0 +1,718 @@
+///
+fn print_data(buffer: &[u8], config: &Args) -> Result<(), Box<dyn std::error::Error>> {
+    let named_chars = get_named_chars(); // Get the named characters for special byte values.
+    let mut offset = 0; // Initialize offset for printing addresses.
+
+    while offset < buffer.len() {
+        let local_buf = &buffer[offset..(offset + 16).min(buffer.len())];
+        // Print the address in the specified base format.
+        if let Some(base) = config.address_base {
+            match base {
+                'd' => print!("{:07} ", offset),
+                'o' => print!("{:07o} ", offset),
+                'x' => print!("{:07x} ", offset),
+                'n' => (),
+                _ => print!("{:07} ", offset),
+            }
+        } else {
+            print!("{:07} ", offset); // Default to octal if no base is specified.
+        }
+
+        if config.bytes_char {
+            let mut previously = String::new();
+            for byte in local_buf {
+                let current = match *byte {
+                    b'\0' => "NUL ".to_string(),
+                    b'\x08' => "BS ".to_string(),
+                    b'\x0C' => "FF ".to_string(),
+                    b'\x0A' => "NL ".to_string(),
+                    b'\x0D' => "CR ".to_string(),
+                    b'\x09' => "HT ".to_string(),
+                    _ if byte.is_ascii_graphic() || byte.is_ascii_whitespace() => {
+                        format!("{} ", *byte as char)
+                    }
+                    _ => format!("{:03o} ", byte),
+                };
+
+                if previously == current && !config.verbose {
+                    print!("* ");
+                    continue;
+                }
+
+                print!("{}", current);
+                previously = current;
+            }
+            println!(); // Print a newline after each line of bytes.
+        } else if config.type_strings.is_empty() {
+            let mut previously = String::new();
+            for byte in local_buf {
+                let current = format!("{:03o} ", byte);
+                if previously == current && !config.verbose {
+                    print!("* ");
+                    continue;
+                }
+
+                print!("{}", current);
+                previously = current;
+            }
+            println!(); // Print a newline after each line of bytes.
+        } else {
+            for type_string in &config.type_strings {
+                // Determine the number of bytes to read for this type.
+                let mut chars = type_string.chars();
+                let type_char = chars.next().unwrap();
+                let num_bytes: usize = chars.as_str().parse().unwrap_or(match type_char {
+                    'd' | 'u' | 'o' | 'x' => 2,
+                    'f' => 4,
+                    _ => 1,
+                });
+
+                let chunks = local_buf.chunks(num_bytes);
+                match type_char {
```

Comment:
There's definitely a better way to do this. On possible idea is to have a generic function depending on the type_char. Have a trait that describes printing and multiple structs (a,c,u,d,x,o,f,default) that implement this trait.

Function body will have skip checks and printing itself, and the trait method would return `current` value.

Of course you can do it your way, just make sure that it's simpler without a ton of repetition.